tools/rc/usResourceCompiler.cpp

```diff
@@ -0,0 +1,523 @@
+  if(archiveEntry.find_last_of("/") != std::string::npos)
+  {
+    AddDirectory(archiveEntry.substr(0,archiveEntry.find_last_of("/")+1));
+  }
+}
+
+void ZipArchive::AddDirectory(const std::string& dirName)
+{
+  assert(dirName[dirName.length() - 1] == '/');
+  if (archivedDirs.insert(dirName).second)
+  {
+    std::clog << "\t new dir entry " << dirName << std::endl;
+    // The directory entry does not yet exist, so add it
+    if (!mz_zip_writer_add_mem(writeArchive.get(), dirName.c_str(), NULL, 0, MZ_NO_COMPRESSION))
+    {
+      throw std::runtime_error("zip add_mem error");
+    }
+  }
+}
+
+ZipArchive::~ZipArchive()
+{
+  assert(writeArchive->m_zip_mode != MZ_ZIP_MODE_INVALID);
+  std::clog << "Finalizing the zip archive ..." << std::endl;
+  std::clog << "Archive has the following files" << std::endl;
+  for (auto fileNameEntry : archivedNames)
+  {
+    std::clog << "\t " << fileNameEntry << std::endl;
+  }
+  std::clog << "and directory entries" << std::endl;
+  for (auto dirEntry : archivedDirs)
+  {
+    std::clog << "\t " << dirEntry << std::endl;
+  }
+  if (mz_zip_writer_finalize_archive(writeArchive.get()) == 0)
+  {
+    assert(writeArchive->m_zip_mode == MZ_ZIP_MODE_WRITING_HAS_BEEN_FINALIZED);
+    if (!mz_zip_writer_end(writeArchive.get()))
+    {
+      std::cerr << "Could not close zip archive writer" << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+void ZipArchive::AddResourcesFromArchive(const std::string &archiveFileName)
+{
+  mz_zip_archive currZipArchive;
+  mz_uint currZipIndex = 0;
+  memset(&currZipArchive, 0, sizeof(mz_zip_archive));
+  std::clog << "Merging zip file " << archiveFileName << " ... " << std::endl;
+  if (!mz_zip_reader_init_file(&currZipArchive, archiveFileName.c_str(), 0))
+  {
+    throw std::runtime_error("Could not initialize zip archive " + archiveFileName);
+  }
+  char archiveName[MZ_ZIP_MAX_ARCHIVE_FILENAME_SIZE];
+  mz_uint numZipIndices = mz_zip_reader_get_num_files(&currZipArchive);
+  for (currZipIndex = 0; currZipIndex < numZipIndices; ++currZipIndex)
+  {
+    mz_uint numBytes = mz_zip_reader_get_filename(&currZipArchive, currZipIndex, archiveName, sizeof archiveName);
+    std::clog << "\tmerging: " << archiveName << " (from " << archiveFileName << ") "<< std::endl;
+    try
+    {
+      if (numBytes > 1)
+      {
+        if (archiveName[numBytes-2] != '/') // The last character is '\0' in the array
+        {
+          if (!archivedNames.insert(archiveName).second)
+          {
+            throw std::runtime_error("Found duplicate file with name " + std::string(archiveName));
+          }
+          if (!mz_zip_writer_add_from_zip_reader(writeArchive.get(), &currZipArchive, currZipIndex))
+          {
+            throw std::runtime_error("Failed to append file " + std::string(archiveName) + "from archive " + archiveFileName);
+          }
+        }
+        else
+        {
+          AddDirectory(std::string(archiveName));
+        }
+      }
+    }
```

Comment:
The previous C implementation kept the archived names sorted (```us_archived_names_sort```). Why was that done and does that need to be carried over into this implementation?